server/nuxeo-nxr-server/src/main/resources/templates/common-base/client/scripts/deleteOrphanTasks.groovy

```diff
@@ -0,0 +1,37 @@
+import org.apache.logging.log4j.LogManager
+import org.nuxeo.ecm.platform.task.Task
+import org.nuxeo.ecm.core.api.IdRef
+import org.nuxeo.ecm.core.api.ScrollResult
+import org.nuxeo.runtime.transaction.TransactionHelper
+
+def log = LogManager.getLogger("deleteOrphanTasks.groovy")
+def nxql = "SELECT * FROM Document WHERE ecm:mixinType = 'Task'"
+def dryRun = true
+def session = Context.getCoreSession()
+def ScrollResult<String> scrollResult = session.scroll(nxql, 500, 60)
+def removedCount = 0
+def taskCount = 0
+
+log.info("Starting deleting orphan tasks...")
+while (scrollResult.hasResults()) {
+    taskCount += scrollResult.getResults().size()
+    def cache = [:]
+    for (id in scrollResult.getResults()) {
+        def task = session.getDocument(new IdRef(id)).getAdapter(Task.class)
+        def workflowRef = new IdRef(task.getProcessId())
+        if (!cache.computeIfAbsent(workflowRef, session.&exists)) {
+            if (dryRun) {
+                log.info("Would remove orphan task: {}", id)
+            } else {
+                log.info("Removing orphan task: {}", id)
+                session.removeDocument(new IdRef(id))
+                removedCount += 1
+            }
+        }
+    }
+    scrollResult = session.scroll(scrollResult.getScrollId())
+    TransactionHelper.commitOrRollbackTransaction()
+    TransactionHelper.startTransaction()
```

Comment:
Shouldn't we re-start the transaction *before* doing a new scroll?

Reply:
This is how it's done in `org.nuxeo.ecm.core.BatchFinderWork.work()` lines 88-90
Shouldn't the scroll be a snapshot ? Then it shouldn't be affected by the transaction ?

Reply:
Tested it with a 1 bucket size and 2 tasks. Seems to be ok but in doubt i swapped the lines